File: bro-to-app/select-filedescriptor/receiver.cc
#include "broker/broker.hh"
#include "broker/bro.hh"
#include <sys/select.h>

using namespace broker;

// demo receiver that registers a callback to the topic "demo/select_fd"

int main() {
    endpoint ep;
    auto subscriber = ep.make_subscriber( {"demo/select_fd"} );
    ep.listen("127.0.0.1", 9999);

    fd_set fds;
    while (true) {
        FD_ZERO(&fds);
        FD_SET(subscriber.fd(), &fds);

        // this will block until we have read-readiness on the file descriptor
        auto result = select(FD_SETSIZE, &fds, nullptr, nullptr, nullptr);
        if (result == -1) {
            std::cout << "boom. this is the end." << std::endl;
            return -1;
        }

        if (FD_ISSET(subscriber.fd(), &fds)) {
            auto msg = subscriber.get(); // we could also subscriber.poll() and handle vector of messages.
            broker::topic topic(std::move(msg.first));
            bro::Event event(std::move(msg.second));
            std::cout << "received on topic: " << topic << "     event name: " << event.name() << "    content: " << event.args() << std::endl;
        }
    }

    return 0;
}

File: bro-to-app/select-filedescriptor/receiver.py
#!/usr/bin/python

import broker
import select

## demo receiver that is subscribed to the topic "demo/select_fd"

ep = broker.Endpoint()
subscriber = ep.make_subscriber("demo/select_fd")
ep.listen("127.0.0.1", 9999)

while(True):

    ## this will block until we have read-readiness on the file descriptor
    # print("wait ...")
    fd_sets = select.select([subscriber.fd()], [], [])
    # print ("go on...")
    if not fd_sets[0]:
        print("boom. this is the end.")

    (topic, data) = subscriber.get() #// we could also subscriber.poll() and handle array of messages
    received_event = broker.bro.Event(data)

    print("received on topic: {}    event name: {}    content: {}".format(topic, received_event.name(), received_event.args()))


## in fact, with a blocking select this is pretty similar to the "simple" example. The call to `subscriber.get()` blocks as well. To handle this nicely, we have to wrap it into a thread.